### PR TITLE
Fixed #28032 -- Added Paginator.get_page().

### PR DESCRIPTION
**Pull request by sjlehtin**
Moved boilerplate from docs to a utility function.
https://code.djangoproject.com/ticket/28032

### CODE REVIEW

File: django/core/paginator.py [hunk 1]
```diff
@@ -96,19 +95,35 @@ def _check_object_list_is_ordered(self):
         """
         Warn if self.object_list is unordered (typically a QuerySet).
         """
-        if hasattr(self.object_list, 'ordered') and not self.object_list.ordered:
+        if hasattr(self.object_list,
+                   'ordered') and not self.object_list.ordered:
```
Comment (Ian-Foote) [inline django/core/paginator.py]: I don't think we want to shorten this line. Django allows [up to 119 characters](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/).

File: django/core/paginator.py [hunk 2]
```diff
@@ -96,19 +95,35 @@ def _check_object_list_is_ordered(self):
...
+        """
+        # This utility method implements the boilerplate code
+        # in previous Django versions documentation.
+        try:
+            number = self.validate_number(number)
+        except PageNotAnInteger:
+            number = 1
+        except EmptyPage:
+            number = self.num_pages
+        return self.page(number)
```
Comment (Ian-Foote) [inline django/core/paginator.py]: It's a bit annoying that we repeat the `self.validate_number` call in `self.page`, but I don't see a nice way to avoid it.

File: django/core/paginator.py [hunk 3]
```diff
@@ -96,19 +95,35 @@ def _check_object_list_is_ordered(self):
...
-        if hasattr(self.object_list, 'ordered') and not self.object_list.ordered:
+        if hasattr(self.object_list,
+                   'ordered') and not self.object_list.ordered:
             warnings.warn(
                 'Pagination may yield inconsistent results with an unordered '
                 'object_list: {!r}'.format(self.object_list),
                 UnorderedObjectListWarning
             )
 
+    def page_force_into_range(self, number):
```
Comment (Ian-Foote) [inline django/core/paginator.py]: I'm not entirely sure about this name. I don't have any better suggestions right now though.

File: django/core/paginator.py (hunk 1 above)
Comment (sjlehtin) [inline django/core/paginator.py]: Ok.

File: django/core/paginator.py (hunk 3 above)
Comment (sjlehtin) [inline django/core/paginator.py]: I'm open to better names.  I just want it to be explicit.

File: django/core/paginator.py (hunk 2 above)
Comment (sjlehtin) [inline django/core/paginator.py]: I had the same thoughts, but this is the least evil I came up with.

**Comment by sjlehtin** [comment]
I'd love to have more suggestions for the name, where should I post a question?

File: django/core/paginator.py [hunk 4]
```diff
@@ -103,12 +102,27 @@ def _check_object_list_is_ordered(self):
...
+        """
+        Always return a valid page, even if the page argument is not in range
+        or is invalid.  If no objects and allow_empty_first_page=False, this
```
Comment (MarkusH) [inline django/core/paginator.py]: Please don't use double spaces here.

File: django/core/paginator.py [hunk 5]
```diff
@@ -23,7 +23,6 @@ class EmptyPage(InvalidPage):
 
 
 class Paginator:
-
```
Comment (MarkusH) [inline django/core/paginator.py]: Please add this line back, it's unrelated to the remaining patch.

File: django/core/paginator.py [hunk 6]
```diff
@@ -103,12 +102,27 @@ def _check_object_list_is_ordered(self):
                 UnorderedObjectListWarning
             )
 
+    def page_force_into_range(self, number):
```
Comment (MarkusH) [inline django/core/paginator.py]: The name for this function is a bit strong, from my perspective. How about `effective_page()`?

File: django/core/paginator.py (hunk 2 above)
Comment (MarkusH) [inline django/core/paginator.py]: Having `self.validate_number` being called twice doesn't matter too much, I think.

File: django/core/paginator.py [hunk 7]
```diff
@@ -103,12 +102,27 @@ def _check_object_list_is_ordered(self):
...
+        """
+        Always return a valid page, even if the page argument is not in range
+        or is invalid.  If no objects and allow_empty_first_page=False, this
+        will still throw an exception.
+        """
+        # This utility method implements the boilerplate code
```
Comment (MarkusH) [inline django/core/paginator.py]: I don't think these 2 lines of comments are necessary.

File: django/core/paginator.py [hunk 8]
```diff
@@ -103,12 +102,27 @@ def _check_object_list_is_ordered(self):
...
+            number = self.num_pages
+        return self.page(number)
+
 
-QuerySetPaginator = Paginator   # For backwards-compatibility.
+QuerySetPaginator = Paginator  # For backwards-compatibility.
 
 
 class Page(collections.Sequence):
-
```
Comment (MarkusH) [inline django/core/paginator.py]: Can you undo this change, please.

File: docs/topics/pagination.txt [hunk 9]
```diff
@@ -104,6 +104,20 @@ The view function looks like this::
 
         return render(request, 'list.html', {'contacts': contacts})
 
+The boilerplate code above is also available as utility function
```
Comment (MarkusH) [inline docs/topics/pagination.txt]: I think I'd just replace the code above with the new API. I don't see a reason why one would wan to do this manually.

File: docs/topics/pagination.txt [hunk 10]
```diff
@@ -182,6 +196,17 @@ Methods
     Returns a :class:`Page` object with the given 1-based index. Raises
     :exc:`InvalidPage` if the given page number doesn't exist.
 
+.. method:: Paginator.page_force_into_range(number)
+
+    .. versionadded:: 2.0
+
+    Returns a :class:`Page` object with the given 1-based index.  If the page
```
Comment (MarkusH) [inline docs/topics/pagination.txt]: Please use "Return" instead of "Return_s_" and don't use double whitespaces after periods.

**Comment by MarkusH** [comment]
Thank you very much for your contribution, @sjlehtin :) Please let me know when you have (more) questions! :+1:

File: django/core/paginator.py (hunk 5 above)
Comment (sjlehtin) [inline django/core/paginator.py]: ok.

File: django/core/paginator.py (hunk 6 above)
Comment (sjlehtin) [inline django/core/paginator.py]: Will do.

File: docs/topics/pagination.txt (hunk 10 above)
Comment (sjlehtin) [inline docs/topics/pagination.txt]: The format "Returns" is used with every other function in the doc file, are you sure?

File: docs/topics/pagination.txt (hunk 9 above)
Comment (sjlehtin) [inline docs/topics/pagination.txt]: Ok.

**Comment by sjlehtin** [comment]
I've addressed everything except the "Returns", I hope.  "effective_page" is a good name, was there prior art for "effective_" in Django somewhere?

**Comment by sjlehtin** [comment]
I now changed the function name in the commit message as well.

File: django/core/paginator.py [hunk 11]
```diff
@@ -103,8 +103,23 @@ def _check_object_list_is_ordered(self):
                 UnorderedObjectListWarning
             )
 
+    def effective_page(self, number):
```
Comment (jschneier) [inline django/core/paginator.py]: Thoughts about a name like `get_page` or something with `safe` in it? I don't think "effective" has much meaning here, it's really about the error handling.

I see that there is already a `_get_page` function (although the fact that it is a hook according to its docstring means it should probably be public). What about putting the logic in there?

Comment (sjlehtin) [inline django/core/paginator.py]: no argument from me for either name.

File: django/core/paginator.py (hunk 2 above)
Comment (schinckel) [inline django/core/paginator.py]: It looks like this is showing up as a requested change, but I don't think that was @MarkusH's intent?

Comment (timgraham) [reply django/core/paginator.py]: I pushed some updates to the branch. I added the `test_get_page_empty_object_list_and_allow_empty_first_page_false` test. This test fails if `validate_number()` isn't called twice so the current approach seems okay to me.
